src/Dependencies/Collections/RoslynImmutableInterlocked.cs

```diff
@@ -0,0 +1,369 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.CodeAnalysis.Collections
+{
+    internal static class RoslynImmutableInterlocked
```

Comment:
>RoslynImmutableInterlocke [](start = 26, length = 25)

I'd move these helpers to `ImmutableSegmentedDictionary` static class. Exposing "RoslynImmutableInterlocked" type is odd.

Reply:
➡️ This type is intentionally designed as a drop-in replacement for code currently using `ImmutableInterlocked` with `ImmutableDictionary<TKey, TValue>`.

Reply:
That may be, but I don't like these type names. You can drop-in replace as well, just use different type names.

Reply:
These names follow the pattern of `RoslynDebug` and `RoslynString`. The methods it contains are exactly the same as the methods in `ImmutableInterlocked` (signatures and semantics), but with signatures adjusted to work with our types in ways that `ImmutableInterlocked` does not support. I can rename this to `ImmutableInterlocked` as soon as C# is updated to allow static extensions to existing static classes.

Reply:
RoslynDebug and RoslynString are internal types limited to Roslyn repo and not exposed by a source package. What if we had another type that needs interlocked operations? Are we gonna have `RoslynImmutableInterlocked2`?

Reply:
How about a nested static class `ImmutableSegmentedDictionary.Interlocked`?

Reply:
How about `ImmutableInterlockedSegmentedDictionary`?

Reply:
I still think `ImmutableSegmentedDictionary` would be reasonable place to put them.

Reply:
> How about `ImmutableInterlockedSegmentedDictionary`?

Other members are going to be added to this type when `ImmutableSegmentedList<T>` is added. `ImmutableInterlocked` provides atomic operations for all of the immutable collections types.

Reply:
> ImmutableInterlocked provides atomic operations for all of the immutable collections types.

That would require either static extension methods or all immutable types defined in the same assembly.

> Other members are going to be added to this type when ImmutableSegmentedList<T> is added

`SegmentedCollectionsInterlocked`?

Reply:
That works for `ImmutableSegmentedDictionary<TKey, TValue>` and `ImmutableSegmentedList<T>`, but not for [`ImmutableArrayDictionary<TKey, TValue>`](#47641).

Reply:
I'll file a follow-up issue to discuss naming for static helper types in source packages.

Reply:
Is the plan to move these to System.Collections.Immutable at some point?

Reply:
Filed #50750. Moving to System.Collections.Immutable is an option, though I'm not sure if/when that would occur.